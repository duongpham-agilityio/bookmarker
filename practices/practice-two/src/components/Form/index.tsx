import { MouseEvent, memo } from 'react';

// Hooks
import { useForm } from 'hooks';

// Components
import { Button, Heading, Input } from 'components/commons';

// Types
import { Book } from 'types';

// Styles
import styles from 'components/Form/index.module.css';

// Assets
import UploadIcon from 'assets/icons/upload.svg';

export type FormProps = {
  value: Omit<Book, 'publishDate' | 'deletedAt' | 'createdAt' | 'updatedAt'> & {
    publishDate?: number;
  };
  title?: string;
  className?: string;
  type?: 'update' | 'create';
  onClose?: (_event: MouseEvent) => void;
};

const Form = (props: FormProps) => {
  const {
    value: data,
    title = 'Create a new book marker',
    className = '',
    type = 'create',
    onClose,
  } = props;
  const {
    value: { author, description, imageURL, name, publishDate, imageName },
    refImage,
    isUpload,
    booksRecommended,
    handleSelectRecommended,
    resetRecommended,
    onChange,
    onSubmit,
  } = useForm(data, type);

  return (
    <section
      className={styles.overlay}
      onClick={(e: MouseEvent) => {
        if (onClose) onClose(e);
      }}
    >
      <form
        className={`${styles.form} ${className}`}
        action="#"
        method="POST"
        onSubmit={onSubmit}
        onClick={(e: MouseEvent) => {
          e.stopPropagation();

          resetRecommended([]);
        }}
      >
        <Heading label={title} className={styles.heading} />

        <section className={styles.content}>
          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Book name
              </label>
              <Input
                value={name}
                placeholder="Book name"
                name="name"
                id="bookName"
                onChange={onChange}
              />

              {!!booksRecommended.length && (
                <ul className={styles.recommend}>
                  {booksRecommended.map((book, index) => {
                    return (
                      <li
                        className={styles.recommendItem}
                        key={index}
                        onClick={(event: MouseEvent) => {
                          event.preventDefault();

                          handleSelectRecommended(index);
                        }}
                      >
                        <p>{book.name}</p>
                      </li>
                    );
                  })}
                </ul>
              )}
            </div>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="author">
                Author
              </label>
              <Input
                value={author}
                placeholder="Author"
                name="author"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="publishedDate">
                Published date
              </label>
              <Input
                type="date"
                value={publishDate}
                name="publishDate"
                onChange={onChange}
              />
            </div>
            <div className={`${styles.formField} ${styles.upload}`}>
              <div className={styles.uploadItem}>
                <label className={styles.label} htmlFor="imageURL">
                  Upload
                </label>
                <input
                  ref={refImage}
                  hidden
                  className={styles.input}
                  type="file"
                  name="imageURL"
                  id="imageURL"
                  accept=".png,.jpeg"
                  onChange={onChange}
                />
                {!imageName ? (
                  <>
                    <Button
                      label="Upload"
                      width="w-lg"
                      border="b-lg"
                      variant="primary"
                      leftIcon={UploadIcon}
                      onClick={() => {
                        refImage.current?.click();
                      }}
                    />
                  </>
                ) : (
                  <p
                    onClick={() => {
                      refImage.current?.click();
                    }}
                  >
                    {imageName}
                  </p>
                )}
              </div>
              <div className={styles.uploadItem}>
                {imageURL && (
                  <img
                    src={imageURL}
                    alt="image book"
                    className={styles.image}
                  />
                )}
              </div>
            </div>
          </div>

          <div className={styles.formItem}>
            <div className={styles.formField}>
              <label className={styles.label} htmlFor="bookName">
                Description
              </label>
              <Input
                variant="area"
                className={styles.text}
                value={description}
                placeholder="Book name"
                name="description"
                onChange={onChange}
              />
            </div>
          </div>

          <div className={styles.action}>
            <Button
              label="Cancel"
              width="w-lg"
              border="b-lg"
              onClick={onClose}
            />
            <Button
              disabled={isUpload}
              type="submit"
              label="Save"
              variant={isUpload ? 'default' : 'primary'}
              width="w-lg"
              border="b-lg"
            />
          </div>
        </section>
      </form>
    </section>
  );
};

export default memo(Form);
