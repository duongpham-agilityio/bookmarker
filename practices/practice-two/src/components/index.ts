export {
  default as Notification,
  type NotificationProps,
} from './Notification';

export { default as Form, type FormProps } from './Form';
export { default as Popup, type PopupProps } from './Popup';
export { default as Card, type CardProps } from './Card';
export { default as CardSkeleton } from './Card/Skeleton';
