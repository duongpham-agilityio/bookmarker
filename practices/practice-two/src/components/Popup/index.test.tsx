import { fireEvent, render } from '@testing-library/react';

// Component
import { Popup } from 'components';

jest.mock('assets/icons/x-letter.svg', () => ({
  default: 'assets/icons/x-letter.svg',
}));

jest.mock('assets/icons/upload.svg', () => ({
  default: 'assets/icons/upload.svg',
}));

jest.mock('assets/icons/next.svg', () => ({
  default: 'assets/icons/next.svg',
}));

jest.mock('assets/icons/trash.svg', () => ({
  default: 'assets/icons/trash.svg',
}));

jest.mock('assets/icons/back.svg', () => ({
  default: 'assets/icons/back.svg',
}));

jest.mock('assets/icons/pencil.svg', () => ({
  default: 'assets/icons/pencil.svg',
}));

describe('Popup', () => {
  it('match snapshot', () => {
    const { asFragment } = render(
      <Popup
        title="Are you sure to deleted this book"
        description="This action can not undo, so please careful with this action"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('render', () => {
    const { getByText } = render(
      <Popup
        title="Are you sure to deleted this book"
        description="This action can not undo, so please careful with this action"
      />
    );

    expect(getByText(/Are you sure to deleted this book/i)).toBeInTheDocument();
  });

  it('action', () => {
    const mockOnAccept = jest.fn();
    const mockOnCancel = jest.fn();
    const { getAllByRole } = render(
      <Popup
        title="Are you sure to deleted this book"
        description="This action can not undo, so please careful with this action"
        onAccept={mockOnAccept}
        onCancel={mockOnCancel}
      />
    );
    const btns = getAllByRole('button');

    fireEvent.click(btns[0]);
    fireEvent.click(btns[1]);

    expect(mockOnAccept).toHaveBeenCalled();
    expect(mockOnCancel).toHaveBeenCalled();
  });
});
