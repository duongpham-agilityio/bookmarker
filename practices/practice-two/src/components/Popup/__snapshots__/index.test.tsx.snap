// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup match snapshot 1`] = `
<DocumentFragment>
  <section>
    <div>
      <h2
        class="undefined undefined "
      >
        Are you sure to deleted this book
      </h2>
      <p>
        This action can not undo, so please careful with this action
      </p>
      <div>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          Cancel
        </button>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          Ok
        </button>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
